### DISCUSSION TITLE
Basic Auth in Nuxt 3

### DISCUSSION THREAD

**Discussion by JensErikChristiansen** [nuxt/nuxt#16245] [Questions] [answered]
Hi! How do I implement basic auth in Nuxt 3?

I tried https://www.npmjs.com/package/nuxt-basic-auth-module, and it works locally, but as soon as I build it for production, it's not used.

The following code does seem to work, but the nuxt docs say "Middleware handlers should not return anything (nor close or respond to the request) and only inspect or extend the request context or throw an error".

server/middleware/basicAuth.ts:
```
export default defineEventHandler((event) => {
    const config = useRuntimeConfig();
    const base64Credentials = event.req.headers.authorization?.split(' ')[1];

    if (base64Credentials) {
        const credentials = Buffer.from(base64Credentials, 'base64').toString(
            'ascii',
        );

        const [username, password] = credentials.split(':');

        if (
            username === config.BASIC_AUTH_USERNAME &&
            password === config.BASIC_AUTH_PASSWORD
        ) {
            return;
        }
    }

    event.res.statusCode = 401;
    event.res.setHeader('WWW-Authenticate', 'Basic realm="blah"');

    return event.res.end();
});
```

Another question is: this code is interfering with SSR. A component that is being rendered on the server is trying to call an api in server/api, but the basic auth is blocking it. Any thoughts?

Thank you in advance!

**Comment by Diizzayy** [comment, marked as answer]
> The following code does seem to work, but the nuxt docs say "Middleware handlers should not return anything (nor close or respond to the request) and only inspect or extend the request context or throw an error"

Indeed, doing so may lead to unintended side effects.

I haven't tested it but the following code should meet your needs.

```ts
export default defineEventHandler((event) => {
    let allow = false;

    const config = useRuntimeConfig();
    const base64Credentials = event.req.headers?.authorization?.split(" ")?.[1];

    if (base64Credentials) {
        const credentials = Buffer.from(base64Credentials, "base64").toString("ascii");

        const [username, password] = credentials.split(":");

        allow = username === config?.BASIC_AUTH_USERNAME && password === config?.BASIC_AUTH_PASSWORD;
    }

    if (!allow) {
        event.res.statusCode = 401;
        event.res.setHeader("WWW-Authenticate", 'Basic realm="blah"');
        event.res.end('Unauthorized');
    }
});
```

> this code is interfering with SSR. A component that is being rendered on the server is trying to call an api in server/api, but the basic auth is blocking it. Any thoughts?

@JensErikChristiansen  The initial request (server rendered) isn't aware of the user's auth credentials, hence it is being blocked by basic auth.

One way to circumvent this is by proxying your browser's authorization header to the server as seen below using [the `useRequestHeaders` composable](http://v3.nuxtjs.org/api/composables/use-request-headers)

```ts
const { data } = await useFetch('/api/private', {
    // proxy browser's `authorization` header to server for SSR
    headers: useRequestHeaders(['authorization'])
})
```

**Reply by JensErikChristiansen** [reply]
`useRequestHeaders(['authorization'])` works, thank you!

The code above that, though, is pretty much the same as mine, no? I understand that we're not returning anything at the bottom of the function, but that doesn't mean that we don't end the request if `!allow`, right?

**Reply by Diizzayy** [reply]
> `useRequestHeaders(['authorization'])` works, thank you!
> 
> The code above that, though, is pretty much the same as mine, no?

@JensErikChristiansen  Precisely, it's your code with just a slight improvement.

> but that doesn't mean that we don't end the request if !allow, right?

Indeed. We end the request for the `!allow` condition, doing so is in fact acceptable.

**Reply by JensErikChristiansen** [reply]
Perfect, thank you so much!
